### PR TITLE
feat(Select): add select component

### PR DESCRIPTION
**Pull request by booc0mtaco**
### Summary:

[EDS-754]: create a new `Select` component, based closely on the
existing `Dropdown` component. Make use of subcomponents introduced
in the `Menu` component, like PopoverListItem and PopoverContainer,
and some light refactoring to simplify component structure for better
alignment with Menu.

- mark appropriate components as deprecated
- merge button trigger component into select
- add specificity to class naming on modifiers
- tidy up API by removing options flagged as unused

### Test Plan:

- [x] storybook tests
- [x] chromatic

[EDS-754]: https://czi-tech.atlassian.net/browse/EDS-754?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/Select/Select.module.css [hunk 1]
```diff
@@ -0,0 +1,115 @@
+@import '../../design-tokens/mixins.css';
+
+/*------------------------------------*\
```
Comment (booc0mtaco) [inline src/components/Select/Select.module.css:3]: Next up is renaming all the classes and such to be `select` based

File: src/components/Select/Select.stories.module.css [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/*------------------------------------*\
```
Comment (booc0mtaco) [inline src/components/Select/Select.stories.module.css:1]: Next up is renaming all the classes to be `select`-based

File: src/components/Select/Select.module.css [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+/**
+ * The caret icon to decorate the select trigger button, animated to rotate.
+ */
+.select-button__icon {
+  transform: rotate(0);
+
+  transition: transform var(--eds-anim-move-medium) ease-out;
+  @media (prefers-reduced-motion) {
```
Comment (jinlee93) [inline src/components/Select/Select.module.css:107]: nice

File: src/components/Select/Select.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  component: Select,
+  parameters: {
+    badges: [BADGE.BETA],
+    layout: 'centered',
+  },
+  decorators: [
+    (Story) => (
+      <div
+        style={{
+          margin: '1rem', // Provides spacing to see activity around the select field
```
Comment (jinlee93) [inline src/components/Select/Select.stories.tsx:22]: If layout is `centered` does this still make a difference?

File: src/components/Select/Select.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+      variant="compact"
+    />
+  ),
+};
+
+export const UsingFunctionChildrenProp: StoryObj = {
+  render: () => <InteractiveExampleUsingFunctionChildren />,
+};
+
+// This story just opens the dropdown automatically so chromatic can test it.
```
Comment (jinlee93) [inline src/components/Select/Select.stories.tsx:208]: 1. You can use [`isChromatic`](https://www.chromatic.com/docs/ischromatic) to add play functions to the above stories if you want to snap the open state similarly (unless ofc your intention was to snap the button)
2. If you don't want to snap, you can always [disable](https://www.chromatic.com/docs/ignoring-elements#ignore-stories) the snaps in Chromatic
3. Something about timing and animations may not snap very nicely so can be flaky in Chromatic, just something to watch out for
  - Did just see a possibly new Chromatic feature regarding pausing [CSS animations](https://www.chromatic.com/docs/animations#css-animations) if you wanted to check that out

File: src/components/Select/Select.test.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        </Select.Options>
+      </Select>
+    );
+    const renderMethod = () => {
+      render(dropdownWithDropdownLabel);
+    };
+
+    expect(renderMethod).not.toThrow(Error);
+  });
+});
```
Comment (jinlee93) [inline src/components/Select/Select.test.tsx:85]: Do you think there is value in testing the keyboard navigation functionality or would we just be testing HeadlessUI functionality that we don't own?

File: src/components/Select/Select.tsx [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+  /**
+   * Optional className for additional options menu styling.
+   *
+   * When not using the compact variant, if optionsClassName is provided please
+   * include the width property to define the options menu width.
+   */
+  optionsClassName?: string;
+};
+
+function childrenHaveLabelComponent(children?: ReactNode): boolean {
```
Comment (jinlee93) [inline src/components/Select/Select.tsx:57]: Would this work with React fragments?

**Comment by jinlee93** [comment]
Think you also need to add
```
export { default as Section } from './components/Section';
```
to `src/index.ts`

File: src/components/Select/Select.stories.tsx (hunk 4 above)
Comment (booc0mtaco) [reply src/components/Select/Select.stories.tsx:22]: Yah, it doesn't seem necessary anymore. I will remove this decorator to simplify things

File: src/components/Select/Select.stories.tsx (hunk 5 above)
Comment (booc0mtaco) [reply src/components/Select/Select.stories.tsx:208]: Oh we may have to do a little bit of a dive into some of this. I will add this flag as I would like one of these to have a snap with it open. 

(this particular test was brought over from the `Dropdown` implementation)

File: src/components/Select/Select.test.tsx (hunk 6 above)
Comment (booc0mtaco) [reply src/components/Select/Select.test.tsx:85]: Yah, the latter. if there are any bugs with the keyboard behavior, we would essentially file a bug with upstream library. Now, if we were adding some custom keyboard handling, we def would want to test that.

File: src/components/Select/Select.tsx (hunk 7 above)
Comment (booc0mtaco) [reply src/components/Select/Select.tsx:57]: Some of the factoring in this came from the existing tests I inherited here. I was trying to change only a little of that code, but these functions obfuscate what is being tested a little bit, in general.

I suspect there is something in react that actually does this (or a better job of it), so was inclined to leave it for a later investigation

File: src/components/Select/Select.tsx [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+}
+
+/**
+ * `import {Select} from "@chanzuckerberg/eds";`
+ *
+ * EDS Select. Used to select one option from a list of options.
+ *
+ * Built on top of the Headless UI Listbox: https://headlessui.dev/react/listbox#basic-example
+ *
+ * TODO: document using class names to alter widths
```
Comment (jinlee93) [inline src/components/Select/Select.tsx:83]: Leftover todo?

**Review by jinlee93 — APPROVED** [review]
Looks good to me but would love to solicit @dierat's opinions

File: src/components/Dropdown/Dropdown.tsx [hunk 9]
```diff
@@ -114,6 +112,10 @@ function childrenHaveLabelComponent(children?: ReactNode): boolean {
 /**
  * `import {Dropdown} from "@chanzuckerberg/eds";`
  *
+ * Note: this component has been deprecated and will be removed in a future release.
+ * Please use the Select component instead for forms, and the Menu component for other
+ * types of popover-like dropdowns.
+ *
```
Comment (dierat) [inline src/components/Dropdown/Dropdown.tsx:118]: Could you add `@deprecated` to the bottom of this component docstring and add the deprecated badge to `Dropdown.stories.tsx`?

File: src/components/Select/Select.tsx [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    ) {
+      return true;
+    } else if ('props' in child && child.props.children) {
+      return childrenHaveLabelComponent(child.props.children);
+    }
+    return false;
+  });
+}
+
+/**
```
Comment (dierat) [inline src/components/Select/Select.tsx:76]:
```suggestion
/**
 * BETA: This component is still a work in progress and is subject to change.
 *
```

**Comment by dierat** [comment]
Looks good!

File: src/components/Select/Select.tsx (hunk 8 above)
Comment (booc0mtaco) [reply src/components/Select/Select.tsx:85]: yah sorta? I wasn't sure how this currently works at first, but I did include a test with the result in, so i can address that

File: src/components/Dropdown/Dropdown.tsx (hunk 9 above)
Comment (dierat) [reply src/components/Dropdown/Dropdown.tsx:118]: Thanks for adding the `@deprecated` tag! Can you also add the badge in `Dropdown.stories.tsx`?

File: src/components/Select/Select.module.css [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  margin-bottom: var(--eds-size-half);
+}
+
+/**
+ * Visually hides the label element if a aria-label is passed,
+ * but element is kept for screen reading accessibility.
+ */
+.u-is-vishidden {
+  @mixin visuallyHide;
+}
```
Comment (dierat) [inline src/components/Select/Select.module.css:39]: Where is this being used?

Comment (booc0mtaco) [reply src/components/Select/Select.module.css:39]: oh yah this was a carry-over from the dropdown component. I can remove this

File: src/components/Dropdown/Dropdown.tsx (hunk 9 above)
Comment (booc0mtaco) [reply src/components/Dropdown/Dropdown.tsx:118]: Yep, this is recorded [here](https://github.com/chanzuckerberg/edu-design-system/blob/next/src/components/Dropdown/Dropdown.stories.tsx#L16) which came in with my other popover ticket

Comment (dierat) [reply src/components/Dropdown/Dropdown.tsx:118]: Awesome, thanks!

File: src/components/Select/Select.module.css (hunk 11 above)
Comment (dierat) [reply src/components/Select/Select.module.css]: Thanks! I really wish we had a linter for this..

**Review by dierat — APPROVED** [review]
Looks great, thanks!

**Comment by booc0mtaco** [comment]
Thanks all!

**Pull request merged**